esphome/components/shtcx/shtcx.cpp

```diff
@@ -0,0 +1,165 @@
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with SHTCx failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+
+  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "Humidity", this->humidity_sensor_);
+}
+float SHTCXComponent::get_setup_priority() const { return setup_priority::DATA; }
+void SHTCXComponent::update() {
+  if (this->status_has_warning()) {
+    ESP_LOGW(TAG, "Retrying to reconnect the sensor.");
+    this->write_command_(SHTCX_COMMAND_SOFT_RESET);
+  }
+  if (this->type_ != SHTCX_TYPE_SHTC1) {
+    this->wake_up();
+  }
+  if (!this->write_command_(SHTCX_COMMAND_POLLING_H)) {
+    this->status_set_warning();
+    return;
+  }
+
+  this->set_timeout(50, [this]() {
+    uint16_t raw_data[2];
+    if (!this->read_data_(raw_data, 2)) {
+      this->status_set_warning();
+      return;
+    }
+
+    float temperature = 175.0f * float(raw_data[0]) / 65536.0f - 45.0f;
+    float humidity = 100.0f * float(raw_data[1]) / 65536.0f;
+
+    ESP_LOGD(TAG, "Got temperature=%.2f°C humidity=%.2f%%", temperature, humidity);
+    if (this->temperature_sensor_ != nullptr)
+      this->temperature_sensor_->publish_state(temperature);
+    if (this->humidity_sensor_ != nullptr)
+      this->humidity_sensor_->publish_state(humidity);
+    this->status_clear_warning();
+    if (this->type_ != SHTCX_TYPE_SHTC1) {
+      this->sleep();
+    }
+  });
+}
+
+bool SHTCXComponent::write_command_(uint16_t command) {
+  // Warning ugly, trick the I2Ccomponent base by setting register to the first 8 bit.
+  return this->write_byte(command >> 8, command & 0xFF);
+}
+
+uint8_t sht_crc(uint8_t data1, uint8_t data2) {
+  uint8_t bit;
+  uint8_t crc = 0xFF;
+
+  crc ^= data1;
+  for (bit = 8; bit > 0; --bit) {
+    if (crc & 0x80)
+      crc = (crc << 1) ^ 0x131;
+    else
+      crc = (crc << 1);
+  }
+
+  crc ^= data2;
+  for (bit = 8; bit > 0; --bit) {
+    if (crc & 0x80)
+      crc = (crc << 1) ^ 0x131;
+    else
+      crc = (crc << 1);
+  }
+
+  return crc;
+}
+
+bool SHTCXComponent::read_data_(uint16_t *data, uint8_t len) {
+  const uint8_t num_bytes = len * 3;
+  auto *buf = new uint8_t[num_bytes];
+
+  if (!this->parent_->raw_receive(this->address_, buf, num_bytes)) {
+    delete[](buf);
+    return false;
+  }
+
+  for (uint8_t i = 0; i < len; i++) {
+    const uint8_t j = 3 * i;
+    uint8_t crc = sht_crc(buf[j], buf[j + 1]);
+    if (crc != buf[j + 2]) {
+      ESP_LOGE(TAG, "CRC8 Checksum invalid! 0x%02X != 0x%02X", buf[j + 2], crc);
+      delete[](buf);
+      return false;
+    }
+    data[i] = (buf[j] << 8) | buf[j + 1];
+  }
+
+  delete[](buf);
+  return true;
+}
+
+void SHTCXComponent::soft_reset() {
+  if (!this->write_command_(SHTCX_COMMAND_SOFT_RESET)) {
+    ESP_LOGE(TAG, "Error sending SHTCX_COMMAND_SOFT_RESET");
+    this->mark_failed();
+    return;
+  }
+  delayMicroseconds(200);
+}
+void SHTCXComponent::sleep() {
+  if (!this->write_command_(SHTCX_COMMAND_SLEEP)) {
```

Comment:
```suggestion
  this->write_command_(SHTCX_COMMAND_SLEEP)
```

The if condition is useless here (returns anyway)

Even better would probably be to return the sucess as a boolean